ls1intum/Artemis
# `Development`: Improve notification performance

### Checklist
#### General


#### Server
#### Client

### Motivation and Context


Improve the performance of loading Artemis by removing unnecessary objects.

### Description

Currently Artemis loads all tutorial groups and conversations that the user is part of only to use their ids for subscribing to updates from them.
These two calls are now unified into one request only retrieving the needed ids.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 2 Students

1. Log in to Artemis
2. Login as both students in different browsers
3. Check that the user gets a popup at the bottom right, when they receive a message or reply while e.g. being on the start page. If this is not the case, check the notification settings, if the notifications are enabled for messages.

2. Login as student and instructor in different browsers
3. Put the student into a tutorial group and check that the student sees a popup when the group is updated or deleted
6. Check that the student gets a notification at the top right in the bell without reloading. There should be no popup. Again, if there is none, check the settings.

### Review Progress






#### Performance Review
#### Code Review
#### Manual Tests

### Test Coverage





#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [notification-sidebar.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4992-3/artifact/shared/Coverage-Report-Client-Tests/app/shared/notification/notification-sidebar/notification-sidebar.component.ts.html) | 87.66% | ✅ |
| [notification.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4992-3/artifact/shared/Coverage-Report-Client-Tests/app/shared/notification/notification.service.ts.html) | 84.37% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [NotificationResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4992-3/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/NotificationResource.html) | 93% | ✅ |

### Screenshots
Settings for testing as well as list of things that should trigger a notification:
![Bildschirmfoto 2023-06-16 um 11 25 17](https://github.com/ls1intum/Artemis/assets/38322605/a46de17e-ae72-482c-be41-798cb0975a90)
